Switch back to getAuthors

adamsilverstein opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Fixes https://github.com/WordPress/gutenberg/issues/26476

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamsilverstein committed:
Switch back to getAuthors

adamsilverstein committed:
Use __unstableGetAuthor.

adamsilverstein committed:
Remove deprecated call.

adamsilverstein committed:
Avoid repeated calls to __unstableGetAuthor.

adamsilverstein committed:
Improfe request in __unstableGetAuthor.

youknowriad commented on packages/core-data/src/resolvers.js:
```diff
@@ -29,14 +29,29 @@ import { ifNotResolved, getNormalizedCommaSeparable } from './utils';
+	const users = yield apiFetch( {
+		path: `/wp/v2/users?who=authors&include=${ id }`,
+	} );
+	yield receiveUserQuery( 'author', users );
```
should the first argument here be "authors" to match the same entity as "getAuthors"?

adamsilverstein replied:
Maybe? I think that results in the fetched author overwriting the authors? I can test.

adamsilverstein replied:
Tested changing this and confirmed using the same key results in the query interfering with the `getAuthors` query, causing errors or unexpected results as the results from the two queries overwrite each other.

youknowriad replied:
Mmm, so if I'm reading  this right `getAuthor(1)` and  `getAuthor(2)` are using the same query key which means if I trigger both concurrently, they'll mess with each other's result.

youknowriad replied:
Yes, I did add dynamic query keys  for both getAuthor and getAuthors, I think it will solve both issues so I reverted the change you did with useMemo

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -21,17 +21,23 @@ function PostAuthor() {
-			const author = getUser( getEditedPostAttribute( 'author' ) );
+			const author =
+				postAuthor ||
+				__unstableGetAuthor( getEditedPostAttribute( 'author' ) )[ 0 ];
```
Actually, why do we need `[ 0 ]` here. getAuthor only returns one author right?

adamsilverstein replied:
Yes, but it is returned as an array with a single element. I can change this so `__unstableGetAuthor` returns a single element which makes more sense based on the naming, does that sound good?

youknowriad commented:
Seems like there's some valid test failures.

adamsilverstein commented:
> Seems like there's some valid test failures.

I'll take a look.

adamsilverstein committed:
Merge branch 'master' into fix/26476

adamsilverstein committed:
Move array extraction into `__unstableGetAuthor` selector.

adamsilverstein committed:
Avoid showing author selector when only one user.

adamsilverstein commented:
> Seems like there's some valid test failures.

@youknowriad One issue was the selector was showing even when the site only had one author. 

I had removed the check for `authors.length <= 1` when switching to `getAuthors` because the `authors` in `check.js` uses the same data store as the selector itself, so when you started searching and got zero results, the `authors` in the check would also be empty and the selector would disappear! 

To alleviate this I added `originalAuthors` behind a memo call so it is only set once. What do you think?

youknowriad commented on packages/core-data/src/selectors.js:
```diff
@@ -56,12 +56,21 @@ export const isRequestingEmbedPreview = createRegistrySelector(
+ */
+export function __unstableGetAuthor( state ) {
+	const authors = getUserQueryResults( state, 'author' );
+	return authors ? authors[ 0 ] : [];
```
It should probably return `null`  if the author was not found.

adamsilverstein replied:
Good point, will change.

youknowriad commented:
> the same data store as the selector itself, so when you started searching and got zero results, the authors in the check would also be empty and the selector would disappear!

I think that's the issue, both selectors use different queries, so they shouldn't share the same "query key".

adamsilverstein commented:
> I think that's the issue, both selectors use different queries, so they shouldn't share the same "query key".

Right - that is why I changed the single author query key to 'author' vs 'authors'

youknowriad committed:
fix getAuthor and getAuthors resolvers

youknowriad committed:
simplification

youknowriad commented:
This is looking good for me, but it could use some testing @adamsilverstein

skorasaurus commented:
Hi, 

I tested this out (last commit - https://github.com/WordPress/gutenberg/pull/26554/commits/f507edbf8e7975a6c38a69d928048c987835213d  ) and the select author dropdown does not appear, even as an admin user. I created another user with the 'editor' role. 

I tried on a test site [wordpress 5.5.3], no other plugins installed except this one.

tellthemachines commented:
Tested this branch on top of WP 5.6-beta4 and can confirm the post author dropdown isn't showing for either editor or admin users. The `PostAuthor` component doesn't seem to be rendering at all now 😕

youknowriad committed:
Fix small selector issues

youknowriad commented:
Yep, i just realized I forgot to fix the query key in the selector.😅

skorasaurus commented:
Thanks for the additional fix; with that last commit, I now confirm that the postauthor component now displays as intended and is keyboard accessible as well.

adamsilverstein commented:
@youknowriad - Great! Works well in my testing.

adamsilverstein reviewed (commented):
👍 Looks good!
